website/src/components/Submission/FormOrUploadWrapper.tsx

```diff
@@ -0,0 +1,149 @@
+import { useEffect, useState, type Dispatch, type FC, type SetStateAction } from 'react';
+
+import type { UploadAction } from './DataUploadForm';
+import type { ColumnMapping } from './FileUpload/ColumnMapping';
+import { SequenceEntryUpload } from './FileUpload/SequenceEntryUploadComponent';
+import type { ProcessedFile } from './FileUpload/fileProcessing';
+import type { InputField } from '../../types/config';
+import type { ReferenceGenomesSequenceNames } from '../../types/referencesGenomes';
+import { EditableMetadata, EditableSequences, MetadataForm, SequencesForm } from '../Edit/InputForm';
+
+export type InputMode = 'form' | 'bulk';
+
+export type SequenceData = {
+    type: 'ok';
+    metadataFile: File;
+    sequenceFile?: File;
+};
+
+export type InputError = {
+    type: 'error';
+    errorMessage: string;
+};
+
+export type FileFactory = () => Promise<SequenceData | InputError>;
+
+type FormOrUploadWrapperProps = {
+    inputMode: InputMode;
+    setFileFactory: Dispatch<SetStateAction<FileFactory | undefined>>;
+    organism: string;
+    action: UploadAction;
+    referenceGenomeSequenceNames: ReferenceGenomesSequenceNames;
+    metadataTemplateFields: Map<string, InputField[]>;
+    enableConsensusSequences: boolean;
+};
+
+/**
+ * A component that allows users to upload sequence data. Two modes are supported, see {@link InputMode}.
+ * In 'form' mode, a form is displayed, and the user can directly enter stuff into the form to upload their
+ * metadata. In 'bulk' mode, the user needs to upload files containing the sequences and metadata.
+ * Either way, the component turns the uploaded data into files, so they can be submitted to the API.
+ * Set the 'fileCreatorSetter' to get the files - have a look at existing usage on how this works.
+ */
+export const FormOrUploadWrapper: FC<FormOrUploadWrapperProps> = ({
+    inputMode,
+    setFileFactory,
+    organism,
+    action,
+    referenceGenomeSequenceNames,
+    metadataTemplateFields,
+    enableConsensusSequences,
+}) => {
+    const isMultiSegmented = referenceGenomeSequenceNames.nucleotideSequences.length > 1;
+    const [editableMetadata, setEditableMetadata] = useState(EditableMetadata.empty());
+    const [editableSequences, setEditableSequences] = useState(
+        EditableSequences.fromSequenceNames(referenceGenomeSequenceNames.nucleotideSequences),
+    );
+
+    const [metadataFile, setMetadataFile] = useState<ProcessedFile | undefined>(undefined);
+    const [sequenceFile, setSequenceFile] = useState<ProcessedFile | undefined>(undefined);
+    // The columnMapping can be null; if null -> don't apply mapping.
+    const [columnMapping, setColumnMapping] = useState<ColumnMapping | null>(null);
+
+    useEffect(() => {
+        setFileFactory(() => {
```

Comment:
I'm not sure I really understand this (which isn't a huge problem). Could you explain a bit about why this works like this (in this thread, not in code necessarily)?

Reply:
explained as a top level comment in the PR